golang/cosmos/types/kv_entry_helpers.go

```diff
@@ -0,0 +1,232 @@
+		// iter.Error() returns an error in both cases
+		return next, io.EOF
+	}
+
+	key := ir.iter.Key()
+	if key == nil {
+		return next, fmt.Errorf("nil key yielded by iterator")
+	}
+
+	value := ir.iter.Value()
+	ir.iter.Next()
+	if value == nil {
+		return NewKVEntryWithNoValue(string(key)), nil
+	} else {
+		return NewKVEntry(string(key), string(value)), nil
+	}
+}
+
+func (ir kvIteratorReader) Close() error {
+	return ir.iter.Close()
+}
+
+var _ KVEntryReader = &kvEntriesReader[any]{}
+
+// kvEntriesReader is the KVEntryReader using an underlying slice of generic
+// kv entries. It reads from the slice sequentially using a type specific
+// toKVEntry func, performing bounds checks, and tracking the position.
+type kvEntriesReader[T any] struct {
+	entries   []T
+	toKVEntry func(T) (KVEntry, error)
+	nextIndex int
+}
+
+// Read yields the next KVEntry from the source
+// Implements KVEntryReader
+func (reader *kvEntriesReader[T]) Read() (next KVEntry, err error) {
+	if reader.entries == nil {
+		return next, fmt.Errorf("reader closed")
+	}
+
+	length := len(reader.entries)
+
+	if reader.nextIndex == length {
+		reader.nextIndex += 1
+		return next, io.EOF
+	}
+
+	if reader.nextIndex > length {
+		return next, fmt.Errorf("index %d is out of source bounds (length %d)", reader.nextIndex, length)
+	}
+
+	entry, err := reader.toKVEntry(reader.entries[reader.nextIndex])
+	reader.nextIndex += 1
+
+	return entry, err
+}
+
+// Close releases the source slice
+// Implements KVEntryReader
+func (reader *kvEntriesReader[any]) Close() error {
+	reader.entries = nil
+	return nil
+}
+
+// NewVstorageDataEntriesReader creates a KVEntryReader backed by a
+// vstorage DataEntry slice
+func NewVstorageDataEntriesReader(vstorageDataEntries []*vstoragetypes.DataEntry) KVEntryReader {
+	return &kvEntriesReader[*vstoragetypes.DataEntry]{
+		entries: vstorageDataEntries,
+		toKVEntry: func(sourceEntry *vstoragetypes.DataEntry) (KVEntry, error) {
+			return NewKVEntry(sourceEntry.Path, sourceEntry.Value), nil
+		},
+	}
+}
+
+// NewSwingStoreExportDataEntriesReader creates a KVEntryReader backed by
+// a SwingStoreExportDataEntry slice
+func NewSwingStoreExportDataEntriesReader(exportDataEntries []*types.SwingStoreExportDataEntry) KVEntryReader {
+	return &kvEntriesReader[*types.SwingStoreExportDataEntry]{
+		entries: exportDataEntries,
+		toKVEntry: func(sourceEntry *types.SwingStoreExportDataEntry) (KVEntry, error) {
+			return NewKVEntry(sourceEntry.Key, sourceEntry.Value), nil
+		},
+	}
+}
+
+// NewJsonRawMessageKVEntriesReader creates a KVEntryReader backed by
+// a json.RawMessage slice
+func NewJsonRawMessageKVEntriesReader(jsonEntries []json.RawMessage) KVEntryReader {
+	return &kvEntriesReader[json.RawMessage]{
+		entries: jsonEntries,
+		toKVEntry: func(sourceEntry json.RawMessage) (entry KVEntry, err error) {
+			err = json.Unmarshal(sourceEntry, &entry)
+			return entry, err
+		},
+	}
+}
+
+var _ KVEntryReader = &jsonlKVEntryDecoderReader{}
```

Comment:
FYI, this var-based check that the structure obeys an interface isn't necessary if the constructor returns the interface type. It's harmless to keep it, though.